R/prepare_historical_xls.R

```diff
@@ -0,0 +1,176 @@
+    #pivot
+    pivot_longer(cols = s1:s20, 
+                 names_to = "stop_num", names_prefix = "s", values_to = "count") %>%
+    mutate(stop_num = as.integer(stop_num)) 
+}
+
+
+#' Corrects species common_names to 
+#' taxonomic standard
+#' @param x a historical .xls
+#' @importFrom stringr str_replace_all
+hist_xls_correct_common_names <- \(x){
+  x <- x %>%
+    mutate(common_name = str_replace(common_name, "Rock Dove", "Rock Pigeon"),
+           common_name = str_replace(common_name, "^Whip-poor-will$", "Eastern Whip-poor-will"))
+}
+
+
+#' Process historical excel sheets
+#' They should all have the same format,
+#' if it is in the wrong format
+#' there will be errors
+#'@importFrom dplyr %>% mutate relocate select matches mutate_all mutate_at
+#'@importFrom readxl read_excel
+#'@importFrom stringr str_extract
+#'@param filenames  a list of historical mbbs .xls files within a folder
+hist_xls_process_xls <- function(filenames) {
+  
+  filename <- str_extract(filenames, "_stops/.*")
+  
+  hist_xls <- 
+    read_excel(filenames,.name_repair = "unique_quiet") %>%
+    hist_xls_filter_to_species_code_rows() %>%
+    hist_xls_rename_columns() %>%
+    mutate(sequence = as.integer(sequence)) %>% #ensure sequence is number
+    dplyr::relocate(common_name, .before = species_code) %>% #readability
+    hist_xls_correct_common_names() %>%
+    hist_xls_flag_missed_species() %>%
+    dplyr::select(common_name, matches("s[0-9]+")) %>% #remove extraneous rows, keep only common name and stops
+    mutate(hist_xls_extract_county_num_year_from_filename(filename),
+           source = "prep_sld, hist xls",
+           route_num = as.integer(route_num)) %>%
+    mutate_all(~ifelse(is.na(.), 0, .)) %>% #change NAs to 0
+    hist_xls_change_xs_to_ones() %>% 
+    #ensure counts are numeric
+    mutate_at(c("s1","s2","s3","s4","s5","s6","s7","s8","s9","s10","s11","s12","s13","s14","s15","s16","s17","s18","s19", "s20"), as.integer) %>%
+    hist_xls_match_mbbs_format()
+  
+  return(hist_xls)
+}
+
+#'Filters to only rows where the first column is a species code
+#'that is four capital letters
+#'or acsp - "Accipiter sp."
+#'@importFrom stringr str_detect
+#'@importFrom dplyr filter
+#'@param x a historical .xls 
+hist_xls_filter_to_species_code_rows <- \(x){
+  x %>%
+    filter(str_detect(.[[1]], "^[A-Z][A-Z][A-Z][A-Z]$|acsp"))
+}
+
+
+#' Rename the columns of a historical xls
+#'@importFrom dplyr %>%
+#'@param x a historical .xls 
+hist_xls_rename_columns <- \(x){
+  x %>%
+    `colnames<-`(c("species_code", "s1","s2","s3","s4","s5","s6","s7","s8","s9","s10","s11","s12","s13","s14","s15","s16","s17","s18","s19", "s20", "sum", "sequence","species_code2", "common_name"))
+}
+
+#' Flags if a species code has been missed
+#' @param df a dataframe that contains the column "sequence"
+hist_xls_flag_missed_species <- function(df){
+  
+  if(nrow(df) != max(df$sequence)) {
+    return("Error! nrows does not match expected number of species codes")
```

Comment:
Do you want this to actually be an error and the function that calls `hist_xls_flag_missed_species` to stop? If so, you can do something simple like:

```r
if(nrow(df) != max(df$sequence)) {
  stop("Error! nrows does not match expected number of species codes")
}
```

Or use [`stopifnot`](https://www.rdocumentation.org/packages/base/versions/3.6.2/topics/stopifnot) (Or something from the [`assertthat` package](https://cran.r-project.org/web/packages/assertthat/readme/README.html) -- which we already use in the package.

Reply:
This is a function that got it's main use in testing, so I debated if I should leave it in or remove it. I think it should be left in case there are more excel sheets that come in and would generate an error here.

Reply:
Ok. Maybe leave a note about it's use case then.